tests/ArgentPonyWarcraftClient.Integration.Tests/TestUtilities/ResilientTestCase.cs

```diff
@@ -0,0 +1,48 @@
+﻿using System;
+using System.ComponentModel;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit.Abstractions;
+using Xunit.Sdk;
+
+namespace ArgentPonyWarcraftClient.Integration.Tests
+{
+    [Serializable]
+    internal class ResilientTestCase : XunitTestCase
+    {
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        [Obsolete("Called by the de-serializer", true)]
+        public ResilientTestCase() { }
+
+        public ResilientTestCase(IMessageSink diagnosticMessageSink, TestMethodDisplay testMethodDisplay, ITestMethod testMethod)
+            : base(diagnosticMessageSink, testMethodDisplay, TestMethodDisplayOptions.None, testMethod, testMethodArguments: null)
+        {
+        }
+
+        public override async Task<RunSummary> RunAsync(IMessageSink diagnosticMessageSink,
+            IMessageBus messageBus,
+            object[] constructorArguments,
+            ExceptionAggregator aggregator,
+            CancellationTokenSource cancellationTokenSource)
+        {
+            const int maxRetries = 3;
+            var runCount = 0;
+
+            while (true)
+            {
+                // Capture and delay messages until we've decided to accept the final result.
+                var resilientMessageBus = new ResilientMessageBus(messageBus);
+
+                var summary = await base.RunAsync(diagnosticMessageSink, resilientMessageBus, constructorArguments, aggregator, cancellationTokenSource);
+                if (aggregator.HasExceptions || summary.Failed == 0 || ++runCount >= maxRetries || !resilientMessageBus.CanRetry)
```

Comment:
This is an important class.  The typical `[Fact]` attributes for xUnit are replaced by `[ResilentFact]` attributes that are wired up to this class for test case execution.  The `ResilientMessageBus` wraps the typical `IMessageBus` so it can queue up messages when retries are required.  It also tests the exception thrown to see if retrying might succeed.